### PR TITLE
Refactor markup render system

### LINKED ISSUE #18592: [Bug Report] Relative links in markdown
# Introduction

When using relative links in markdown files you have to use either `./file.md` or `file.md`. When trying to use `/file.md` you will get redirected to the root of the gitea instance.

# Example

https://try.gitea.io/schorsch/relative-links-in-markdown

image

# Proposal

Change `/file.md` from directing to the web root to the project root

# Credits

The bug was discovered by the codeberg user [ivan-paleo](https://codeberg.org/ivan-paleo). For further information have a look at the issue at codeberg: https://codeberg.org/Codeberg/Community/issues/252

### LINKED ISSUE #25632: Relative image link on different branch broken for github migrations / mirrors
### Description

When migrating a repository from GitHub to Gitea. Links to an image in the README of the main branch that reference an image located on another branch don't link correctly on Gitea.

For Example: https://github.com/hlissner/dotfiles references images in the README that are located on the [screenshots branch](https://github.com/hlissner/dotfiles/tree/screenshots). After the migration they don't render correctly on the Gitea repository. See https://try.gitea.io/oonguphuu5og3OovishooVie/master.

The failures in the log during the migration:
```
...eb/routing/logger.go:102:func1() [I] router: completed GET /REDACTED/master for REDACTED, 200 OK in 71.8ms @ repo/view.go:710(repo.Home)
...eb/routing/logger.go:102:func1() [I] router: completed GET /screenshots/alucard/fakebusy.png for REDACTED, 404 Not Found in 1.5ms @ web/goget.go:20(web.goGet)
...eb/routing/logger.go:102:func1() [I] router: completed GET /screenshots/alucard/desktop.png for REDACTED, 404 Not Found in 1.6ms @ web/goget.go:20(web.goGet)
...eb/routing/logger.go:102:func1() [I] router: completed GET /screenshots/alucard/rofi.png for REDACTED, 404 Not Found in 1.7ms @ web/goget.go:20(web.goGet)
...eb/routing/logger.go:102:func1() [I] router: completed GET /screenshots/alucard/tiling.png for REDACTED, 404 Not Found in 1.8ms @ web/goget.go:20(web.goGet)
```

### Gitea Version

1.20.0+rc2

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image

### Git Version

Try Gitea version

### Operating System

Docker

### How are you running Gitea?

https://try.gitea.io/ and Docker version

### Database

SQLite

### LINKED ISSUE #31411: Incorrect attachment path in the activity markdown when sub-path is used
`https://domain.name/<subpath>/attachments/guid`

image

### LINKED ISSUE #32608: Code syntax highlight render failed
### Description

Code syntax highlight render failed

### Gitea Version

a3881ffa3d90f3f017024f701ca8ef420d492224

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image

image

### Git Version

a3881ffa3d90f3f017024f701ca8ef420d492224

### Operating System

win11

### How are you running Gitea?

windows + ms sql server

### Database

MSSQL

### PR DESCRIPTION
**Pull request by wxiaoguang**
This PR removes (almost) all path tricks, and introduces "renderhelper" package.

Now we can clearly see the rendering behaviors for comment/file/wiki, more details are in "renderhelper" tests.

Fix #31411 , fix  #18592, fix #25632 and maybe more problems. (ps: fix #32608 by the way)

### CODE REVIEW

**Comment by lunny** [comment]
Move the render under `models` seems wired. Since only `repo_model.Repository` has been dependent. Maybe we can create a `repository` interface and move these render codes to `modules/renderer` package? We can also extend `gitrepo.Repository` interface to have more methods.

**Comment by wxiaoguang** [comment]
> Move the render under `models` seems wired. Since only `repo_model.Repository` has been dependent. Maybe we can create a `repository` interface and move these render codes to `modules/renderer` package? We can also extend `gitrepo.Repository` interface to have more methods.

There are enough comments, that's the only choice due to the legacy problematic code structure.

image

File: modules/markup/html.go [hunk 1]
```diff
@@ -260,7 +260,6 @@ func RenderEmoji(ctx *RenderContext, content string) (string, error) {
 }
 
 func postProcess(ctx *RenderContext, procs []processor, input io.Reader, output io.Writer) error {
-	defer ctx.Cancel()
```
Comment (lunny) [inline modules/markup/html.go:263]: Why was this line removed?

Comment (wxiaoguang) [reply modules/markup/html.go:263]: there is no such method anymore.

File: modules/markup/html_test.go [hunk 2]
```diff
@@ -374,188 +338,166 @@ func TestRender_ShortLinks(t *testing.T) {
...
-func TestRender_RelativeMedias(t *testing.T) {
-	render := func(input string, isWiki bool, links markup.Links) string {
-		buffer, err := markdown.RenderString(markup.NewTestRenderContext(links, util.Iif(isWiki, localWikiMetas, localMetas)), input)
-		assert.NoError(t, err)
-		return strings.TrimSpace(string(buffer))
-	}
-
-	out := render(`<img src="LINK">`, false, markup.Links{Base: "/test-owner/test-repo"})
-	assert.Equal(t, `<a href="/test-owner/test-repo/LINK" target="_blank" rel="nofollow noopener"><img src="/test-owner/test-repo/LINK"/></a>`, out)
+/*
```
Comment (lunny) [inline modules/markup/html_test.go]: Missing test?

Comment (wxiaoguang) [reply modules/markup/html_test.go]: No, these tests are wrong.

Comment (wxiaoguang) [reply modules/markup/html_test.go]: Removed the dead code in ef1af4aff50f18411eef1b1510ed31f8a9bd3e2c

The right tests are in `renderhelper/*_test.go`

**Review by lunny — APPROVED** [review]

**Review by BLumia — APPROVED** [review]
